updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,218 @@
+@workflow.defn
+class ClusterManagerWorkflow:
+    def __init__(self) -> None:
+        self.state = ClusterManagerState()
+        # Protects workflow state from interleaved access
+        self.nodes_lock = asyncio.Lock()
+
+    @workflow.signal
+    async def start_cluster(self):
+        self.state.cluster_started = True
+        self.state.nodes = {str(k): None for k in range(25)}
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        self.state.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(
+        self, input: ClusterManagerAllocateNNodesToJobInput
+    ) -> List[str]:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        if self.state.cluster_shutdown:
+            # If you want the client to receive a failure, either add an update validator and throw the
+            # exception from there, or raise an ApplicationError. Other exceptions in the main handler
+            # will cause the workflow to keep retrying and get it stuck.
+            raise ApplicationError(
+                "Cannot allocate nodes to a job: Cluster is already shut down"
+            )
+
+        async with self.nodes_lock:
+            unassigned_nodes = [k for k, v in self.state.nodes.items() if v is None]
+            if len(unassigned_nodes) < input.num_nodes:
+                # If you want the client to receive a failure, either add an update validator and throw the
+                # exception from there, or raise an ApplicationError. Other exceptions in the main handler
+                # will cause the workflow to keep retrying and get it stuck.
+                raise ApplicationError(
+                    f"Cannot allocate {input.num_nodes} nodes; have only {len(unassigned_nodes)} available"
+                )
+            assigned_nodes = unassigned_nodes[: input.num_nodes]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            await self._allocate_nodes_to_job(assigned_nodes, input.task_name)
+            self.state.max_assigned_nodes = max(
+                self.state.max_assigned_nodes,
+                len([k for k, v in self.state.nodes.items() if v is not None]),
+            )
+            return assigned_nodes
+
+    async def _allocate_nodes_to_job(self, assigned_nodes: List[str], task_name: str):
+        await workflow.execute_activity(
+            allocate_nodes_to_job,
+            AllocateNodesToJobInput(nodes=assigned_nodes, task_name=task_name),
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        for node in assigned_nodes:
+            self.state.nodes[node] = task_name
+
+    @workflow.update
+    async def delete_job(self, input: ClusterManagerDeleteJobInput):
```

Comment:
This doesn't return anything, so I think readers will be wondering why it's an update rather than a signal.

Reply:
Updates that don't return anything are totally reasonable (they can still raise errors for instance and just waiting on their completion means you know it completed, both of which are improvements over signals). However, I would strongly recommend a `-> None` type hint here.